engine/server/buildkitcontroller.go

```diff
@@ -94,7 +99,41 @@ func NewBuildkitController(opts BuildkitControllerOpts) (*BuildkitController, er
 
 	genericSolver := solver.NewSolver(solver.SolverOpt{
 		ResolveOpFunc: func(vtx solver.Vertex, builder solver.Builder) (solver.Op, error) {
-			return w.ResolveOp(vtx, llbSolver.Bridge(builder), opts.SessionManager)
+			// if this is an ExecOp, pass in an executor configured with the server ID as found
+			// in the job keys
```

Comment:
This is a neat trick :tada: I hadn't realized we could plug in `ResolveOp` like this, I like it :laughing:

Reply:
Haha yeah, you actually *have* to provide this callback, so we're even just using things as intended now!